### PR TITLE
Solve heat before update phases

### PR DESCRIPTION
**Pull request by nmuch**
In preparation for #239
>  [  ] extend update_phases by a const VectorType* temperature=nullptr and use the same interpolation routines as in heat_transfer_operator; rename src to level_set_as_heaviside @nmuch

this PR moves solving the heat or melt_pool operation before `update_phases()`.

### CODE REVIEW

File: source/flow/two_phase_flow_problem.cpp [hunk 1]
```diff
@@ -70,9 +71,20 @@ namespace MeltPoolDG::Flow
...
+                                       level_set_operation.get_level_set_as_heaviside(),
+                                       dt);
+          }
+
+        // update the two phases
+        update_phases(level_set_operation.get_level_set_as_heaviside(), base_in->parameters);
 
         // accumulate forces: a) gravity force
-        compute_gravity_force(vel_force_rhs, base_in->parameters.base.gravity, true);
+        compute_gravity_force(vel_force_rhs, base_in->parameters.base.gravity, false);
```
Comment (peterrum) [inline source/flow/two_phase_flow_problem.cpp:87]:
```suggestion
        compute_gravity_force(vel_force_rhs, base_in->parameters.base.gravity, !melt_pool_operation);
```
?

Comment (nmuch) [reply source/flow/two_phase_flow_problem.cpp:87]: That's possible too :+1: 

To get rid of

Comment (nmuch) [reply source/flow/two_phase_flow_problem.cpp:87]: That's possible too :+1: 

To get rid of `scratch_data->initialize_dof_vector(vel_force_rhs, vel_dof_idx);` I can add a `zero_out` bool to
https://github.com/MeltPoolDG/MeltPoolDG/blob/c25ad42dbfbfbc04497a03787e2d789ea1cff355/source/melt_pool/melt_pool_operation.cpp#L137-L139

File: source/melt_pool/melt_pool_operation.cpp [hunk 2]
```diff
@@ -195,11 +196,10 @@ namespace MeltPoolDG::MeltPool
...
     if (recoil_pressure_operation)
-      recoil_pressure_operation->compute_recoil_pressure_force(
-        vel_force_rhs,
-        level_set_as_heaviside,
-        heat_operation->get_temperature(),
-        false /*false means add to force vector*/);
+      recoil_pressure_operation->compute_recoil_pressure_force(vel_force_rhs,
+                                                               level_set_as_heaviside,
+                                                               heat_operation->get_temperature(),
+                                                               zero_out);
```
Comment (peterrum) [inline source/melt_pool/melt_pool_operation.cpp:202]: In the else part, you need to zero out the vector if `zero_out` is set, don't you?

**Comment by mschreter** [comment]
@nmuch: Thanks for the changes. Could you please have a look at #253, rebase and apply the relevant changes to the new structure? I hope, the algorithm is now more understandable. #253 also resolves most of my tasks of #239.

File: source/melt_pool/melt_pool_operation.cpp (hunk 2 above)
Comment (nmuch) [reply source/melt_pool/melt_pool_operation.cpp]: You're right. But thanks toe #253 this is not an issue anymore.

**Review by mschreter — APPROVED** [review]
Thank you @nmuch, looks good to me. Once CI has passed, I am merging.

**Comment by mschreter** [comment]
@nmuch: Almost done. Could you please update the test output of:

```
The following tests FAILED:
	 17 - tests/recoil_pressure_simplex.mpirun=4.debug (Failed)
```

Thanks!

**Comment by nmuch** [comment]
> @nmuch: Almost done. Could you please update the test output of:
> 
> ```
> The following tests FAILED:
> 	 17 - tests/recoil_pressure_simplex.mpirun=4.debug (Failed)
> ```
> 
> Thanks!

Oh I missed that one because I don't have metis support on.. I simply moved the `two phase flow: total mass` output line to the correct spot. I should pass now, let's see.

**Pull request merged**